core/src/solution_submission.rs

```diff
@@ -215,29 +222,60 @@ async fn retry_with_gas_price_increase(
                 solution.clone(),
                 claimed_objective_value,
                 gas_price,
-                if gas_price == gas_cap {
-                    None
-                } else {
-                    Some(BLOCK_TIMEOUT)
-                },
+                Some(BLOCK_TIMEOUT),
             )
             .await;
 
-        let gas_price_can_still_increase =
-            gas_price.as_u128() as f64 * MIN_GAS_PRICE_INCREASE_FACTOR < gas_cap.as_u128() as f64;
+        let before_deadline = deadline
+            .map(|deadline| Instant::now() < deadline)
+            .unwrap_or(true);
+        let gas_price_can_still_increase = gas_price.as_u128() as f64
+            * MIN_GAS_PRICE_INCREASE_FACTOR
+            < gas_price_cap.as_u128() as f64;
+
         // Breaking condition for our loop
-        gas_price_can_still_increase
-            && matches!(
-                result,
-                Err(MethodError {
-                    inner: ExecutionError::ConfirmTimeout,
-                    ..
-                })
-            )
+        match result {
+            Err(MethodError {
+                inner: ExecutionError::ConfirmTimeout,
+                ..
+            }) => {
+                if gas_price_can_still_increase && before_deadline {
+                    true
+                } else {
+                    let gas_price = if gas_price_can_still_increase {
+                        gas_price_cap
+                    } else {
+                        U256::from(
+                            (gas_price.as_u128() as f64 * MIN_GAS_PRICE_INCREASE_FACTOR) as u128,
+                        )
+                    };
+                    log::info!(
+                        "trying to cancel solution submission transaction for batch {}",
+                        batch_index
+                    );
```

Comment:
More logging than necessary maybe but I'd like to have this until we see this working correctly at least once.

Reply:
I think this logging is fine. Its not like this path should happen often, and when it does its nice to have lots of logs in order to determine the exact sequence that was executed.

If this were a common path that would be hit multiple times, then maybe the extra log messages could pollute the logs a bit, but I don't think its the case here.